File: apps/web/src/server/assistant/agent-runtime/state.ts
import { newId, newRunId } from "./ids"
import type {
    AgentEvidence,
    AgentObservation,
    AgentPlanStep,
    AgentPlanStepStatus,
    AgentState,
    AgentStopReason,
    AgentTokenUsage,
    TaskComplexity,
} from "./types"

/**
 * Agent 状态容器（§6）。
 *
 * 设计约束：
 * - 完全可序列化（snapshot / restore），不依赖 LLM chat history 就能恢复；
 * - 每次 Tool Call、Skill Load、Delegation 之后都必须更新；
 * - 为后续 Resume / 长任务预留持久化能力，但当前版本不做后台异步执行。
 */
export class AgentStateStore {
    private state: AgentState

    constructor(init: {
        runId?: string
        conversationId: string
        userId: string
        goal: string
        complexity?: TaskComplexity
        now?: number
    }) {
        const now = init.now ?? Date.now()
        this.state = {
            runId: init.runId ?? newRunId(),
            conversationId: init.conversationId,
            userId: init.userId,
            goal: init.goal,
            complexity: init.complexity ?? "simple",
            plan: [],
            completedSteps: [],
            pendingSteps: [],
            loadedSkills: [],
            observations: [],
            evidence: [],
            openQuestions: [],
            assumptions: [],
            toolCallCount: 0,
            delegationCount: 0,
            iteration: 0,
            tokenUsage: { input: 0, output: 0, total: 0 },
            startedAt: now,
            updatedAt: now,
            status: "running",
        }
    }

    /** 从快照恢复（为 Resume 预留） */
    static restore(snapshot: AgentState): AgentStateStore {
        const store = new AgentStateStore({
            runId: snapshot.runId,
            conversationId: snapshot.conversationId,
            userId: snapshot.userId,
            goal: snapshot.goal,
            complexity: snapshot.complexity,
            now: snapshot.startedAt,
        })
        store.state = structuredClone(snapshot)
        return store
    }

    /** 只读引用，供 ToolExecutionContext 传递；外部禁止直接改字段 */
    get current(): AgentState {
        return this.state
    }

    snapshot(): AgentState {
        return structuredClone(this.state)
    }

    private touch(): void {
        this.state.updatedAt = Date.now()
    }

    // ---------------------------------------------------------------- 复杂度
    setComplexity(complexity: TaskComplexity): void {
        this.state.complexity = complexity
        this.touch()
    }

    // -------------------------------------------------------------------- Plan
    /** 生成计划；简单请求不应调用（§7） */
    setPlan(steps: Array<{ id?: string; goal: string; dependsOn?: string[] }>): AgentPlanStep[] {
        this.state.plan = steps.map((step) => ({
            id: step.id ?? newId("step"),
            goal: step.goal,
            status: "pending" as AgentPlanStepStatus,
            ...(step.dependsOn?.length ? { dependsOn: step.dependsOn } : {}),
        }))
        this.syncPlanBuckets()
        this.touch()
        return this.state.plan
    }

    addPlanStep(step: { goal: string; dependsOn?: string[]; afterId?: string }): AgentPlanStep {
        const next: AgentPlanStep = {
            id: newId("step"),
            goal: step.goal,
            status: "pending",
            ...(step.dependsOn?.length ? { dependsOn: step.dependsOn } : {}),
        }
        const index = step.afterId ? this.state.plan.findIndex((item) => item.id === step.afterId) : -1
        if (index >= 0) this.state.plan.splice(index + 1, 0, next)
        else this.state.plan.push(next)
        this.syncPlanBuckets()
        this.touch()
        return next
    }

    updatePlanStep(
        id: string,
        patch: { goal?: string; status?: AgentPlanStepStatus; resultSummary?: string },
    ): AgentPlanStep | null {
        const step = this.state.plan.find((item) => item.id === id)
        if (!step) return null
        if (patch.goal != null) step.goal = patch.goal
        if (patch.status != null) step.status = patch.status
        if (patch.resultSummary != null) step.resultSummary = patch.resultSummary
        this.syncPlanBuckets()
        this.touch()
        return step
    }

    removePlanStep(id: string): boolean {
        const before = this.state.plan.length
        this.state.plan = this.state.plan.filter((item) => item.id !== id)
        const removed = this.state.plan.length !== before
        if (removed) {
            this.syncPlanBuckets()
            this.touch()
        }
        return removed
    }

    /** 重新排序：给定完整 id 顺序，未列出的保持原相对次序追加在后 */
    reorderPlan(orderedIds: string[]): AgentPlanStep[] {
        const byId = new Map(this.state.plan.map((step) => [step.id, step]))
        const next: AgentPlanStep[] = []
        for (const id of orderedIds) {
            const step = byId.get(id)
            if (step) {
                next.push(step)
                byId.delete(id)
            }
        }
        for (const step of this.state.plan) {
            if (byId.has(step.id)) next.push(step)
        }
        this.state.plan = next
        this.syncPlanBuckets()
        this.touch()
        return next
    }

    private syncPlanBuckets(): void {
        this.state.completedSteps = this.state.plan.filter(
            (step) => step.status === "completed" || step.status === "skipped",
        )
        this.state.pendingSteps = this.state.plan.filter(
            (step) => step.status === "pending" || step.status === "running",
        )
    }

    // ------------------------------------------------------------------ Skill
    markSkillLoaded(skillId: string): boolean {
        if (this.state.loadedSkills.includes(skillId)) return false
        this.state.loadedSkills.push(skillId)
        this.touch()
        return true
    }

    hasSkill(skillId: string): boolean {
        return this.state.loadedSkills.includes(skillId)
    }

    // ------------------------------------------------------- Observation/证据
    addObservation(observation: AgentObservation): AgentObservation {
        this.state.observations.push(observation)
        this.touch()
        return observation
    }

    addEvidence(evidence: AgentEvidence[]): void {
        if (evidence.length === 0) return
        // EvidenceStore 去重后可能返回已存在的条目，这里按 id 再挡一次，避免 State 里出现重复
        const known = new Set(this.state.evidence.map((item) => item.id))
        const next = evidence.filter((item) => !known.has(item.id))
        if (next.length === 0) return
        this.state.evidence.push(...next)
        this.touch()
    }

    // ------------------------------------------------------------------ 计数
    incrementToolCall(): number {
        this.state.toolCallCount += 1
        this.touch()
        return this.state.toolCallCount
    }

    incrementDelegation(): number {
        this.state.delegationCount += 1
        this.touch()
        return this.state.delegationCount
    }

    incrementIteration(): number {
        this.state.iteration += 1
        this.touch()
        return this.state.iteration
    }

    addTokenUsage(usage: Partial<AgentTokenUsage>): void {
        const input = usage.input ?? 0
        const output = usage.output ?? 0
        this.state.tokenUsage.input += input
        this.state.tokenUsage.output += output
        this.state.tokenUsage.total += usage.total ?? input + output
        this.touch()
    }

    // ---------------------------------------------------------- 未决问题/假设
    addOpenQuestion(question: string): void {
        if (!question.trim() || this.state.openQuestions.includes(question)) return
        this.state.openQuestions.push(question)
        this.touch()
    }

    addAssumption(assumption: string): void {
        if (!assumption.trim() || this.state.assumptions.includes(assumption)) return
        this.state.assumptions.push(assumption)
        this.touch()
    }

    // ------------------------------------------------------------------ 终态
    finish(status: Exclude<AgentState["status"], "running">, reason?: AgentStopReason): void {
        this.state.status = status
        if (reason) this.state.stopReason = reason
        this.touch()
    }
}

File: apps/api/internal/assistantsvc/runtime/state.go
package runtime

// AgentStateStore Agent 状态容器（对照 state.ts）。
//
// 设计约束：
// - 完全可序列化，不依赖 LLM chat history 就能恢复；
// - 每次 Tool Call、Skill Load 之后都必须更新。
type AgentStateStore struct {
	state *AgentState
}

// NewAgentStateStore 初始化状态。
func NewAgentStateStore(runID, conversationID, userID, goal string, complexity TaskComplexity, now int64) *AgentStateStore {
	if now <= 0 {
		now = nowMs()
	}
	if complexity == "" {
		complexity = ComplexitySimple
	}
	return &AgentStateStore{state: &AgentState{
		RunID:          runID,
		ConversationID: conversationID,
		UserID:         userID,
		Goal:           goal,
		Complexity:     complexity,
		Plan:           []AgentPlanStep{},
		CompletedSteps: []AgentPlanStep{},
		PendingSteps:   []AgentPlanStep{},
		LoadedSkills:   []string{},
		Observations:   []AgentObservation{},
		Evidence:       []AgentEvidence{},
		OpenQuestions:  []string{},
		Assumptions:    []string{},
		TokenUsage:     AgentTokenUsage{},
		StartedAt:      now,
		UpdatedAt:      now,
		Status:         StatusRunning,
	}}
}

func (s *AgentStateStore) touch() { s.state.UpdatedAt = nowMs() }

// Current 只读引用，供 ToolExecutionContext 传递；外部禁止直接改字段。
func (s *AgentStateStore) Current() *AgentState { return s.state }

// Snapshot 深拷贝快照。
func (s *AgentStateStore) Snapshot() *AgentState { return cloneState(s.state) }

func (s *AgentStateStore) setComplexity(complexity TaskComplexity) {
	s.state.Complexity = complexity
	s.touch()
}

// SetPlan 生成计划；简单请求不应调用。
func (s *AgentStateStore) SetPlan(drafts []PlanStepDraft) []AgentPlanStep {
	plan := make([]AgentPlanStep, 0, len(drafts))
	for _, d := range drafts {
		plan = append(plan, AgentPlanStep{ID: NewID("step"), Goal: d.Goal, Status: PlanPending, DependsOn: d.DependsOn})
	}
	s.state.Plan = plan
	s.syncPlanBuckets()
	s.touch()
	return s.state.Plan
}

// AddPlanStep 追加步骤（可指定 afterId）。
func (s *AgentStateStore) AddPlanStep(goal string, dependsOn []string, afterID string) AgentPlanStep {
	next := AgentPlanStep{ID: NewID("step"), Goal: goal, Status: PlanPending, DependsOn: dependsOn}
	index := -1
	if afterID != "" {
		for i := range s.state.Plan {
			if s.state.Plan[i].ID == afterID {
				index = i
				break
			}
		}
	}
	if index >= 0 && index+1 <= len(s.state.Plan) {
		tail := append([]AgentPlanStep{}, s.state.Plan[index+1:]...)
		s.state.Plan = append(append(s.state.Plan[:index+1], next), tail...)
	} else {
		s.state.Plan = append(s.state.Plan, next)
	}
	s.syncPlanBuckets()
	s.touch()
	return next
}

// UpdatePlanStep 更新步骤；未命中返回 false。
func (s *AgentStateStore) UpdatePlanStep(id string, goal string, status *AgentPlanStepStatus, resultSummary string) bool {
	for i := range s.state.Plan {
		step := &s.state.Plan[i]
		if step.ID != id {
			continue
		}
		if goal != "" {
			step.Goal = goal
		}
		if status != nil {
			step.Status = *status
		}
		if resultSummary != "" {
			step.ResultSummary = resultSummary
		}
		s.syncPlanBuckets()
		s.touch()
		return true
	}
	return false
}

// RemovePlanStep 删除步骤。
func (s *AgentStateStore) RemovePlanStep(id string) bool {
	kept := s.state.Plan[:0]
	removed := false
	for _, step := range s.state.Plan {
		if step.ID == id {
			removed = true
			continue
		}
		kept = append(kept, step)
	}
	s.state.Plan = kept
	if removed {
		s.syncPlanBuckets()
		s.touch()
	}
	return removed
}

// ReorderPlan 重新排序：给定完整 id 顺序，未列出的保持原相对次序追加在后。
func (s *AgentStateStore) ReorderPlan(orderedIDs []string) []AgentPlanStep {
	byID := map[string]AgentPlanStep{}
	for _, step := range s.state.Plan {
		byID[step.ID] = step
	}
	next := make([]AgentPlanStep, 0, len(s.state.Plan))
	seen := map[string]bool{}
	for _, id := range orderedIDs {
		if step, ok := byID[id]; ok {
			next = append(next, step)
			seen[id] = true
		}
	}
	for _, step := range s.state.Plan {
		if !seen[step.ID] {
			next = append(next, step)
		}
	}
	s.state.Plan = next
	s.syncPlanBuckets()
	s.touch()
	return s.state.Plan
}

func (s *AgentStateStore) syncPlanBuckets() {
	completed := []AgentPlanStep{}
	pending := []AgentPlanStep{}
	for _, step := range s.state.Plan {
		if step.Status == PlanCompleted || step.Status == PlanSkipped {
			completed = append(completed, step)
		} else if step.Status == PlanPending || step.Status == PlanRunning {
			pending = append(pending, step)
		}
	}
	s.state.CompletedSteps = completed
	s.state.PendingSteps = pending
}

// MarkSkillLoaded 标记技能已加载；重复返回 false。
func (s *AgentStateStore) MarkSkillLoaded(skillID string) bool {
	for _, id := range s.state.LoadedSkills {
		if id == skillID {
			return false
		}
	}
	s.state.LoadedSkills = append(s.state.LoadedSkills, skillID)
	s.touch()
	return true
}

// HasSkill 技能是否已加载。
func (s *AgentStateStore) HasSkill(skillID string) bool {
	for _, id := range s.state.LoadedSkills {
		if id == skillID {
			return true
		}
	}
	return false
}

// AddObservation 记录观察。
func (s *AgentStateStore) AddObservation(observation AgentObservation) {
	s.state.Observations = append(s.state.Observations, observation)
	s.touch()
}

// AddEvidence 记录证据（按 id 去重）。
func (s *AgentStateStore) AddEvidence(evidence []AgentEvidence) {
	if len(evidence) == 0 {
		return
	}
	known := map[string]bool{}
	for _, item := range s.state.Evidence {
		known[item.ID] = true
	}
	for _, item := range evidence {
		if known[item.ID] {
			continue
		}
		known[item.ID] = true
		s.state.Evidence = append(s.state.Evidence, item)
	}
	s.touch()
}

// IncrementToolCall 工具调用计数 +1。
func (s *AgentStateStore) IncrementToolCall() int {
	s.state.ToolCallCount++
	s.touch()
	return s.state.ToolCallCount
}

// IncrementDelegation 委派计数 +1。
func (s *AgentStateStore) IncrementDelegation() int {
	s.state.DelegationCount++
	s.touch()
	return s.state.DelegationCount
}

// IncrementIteration 迭代计数 +1。
func (s *AgentStateStore) IncrementIteration() int {
	s.state.Iteration++
	s.touch()
	return s.state.Iteration
}

// AddTokenUsage 累加 token 用量。
func (s *AgentStateStore) AddTokenUsage(input, output int64) {
	s.state.TokenUsage.Input += input
	s.state.TokenUsage.Output += output
	total := input + output
	s.state.TokenUsage.Total += total
	s.touch()
}

// AddOpenQuestion 登记待解决问题。
func (s *AgentStateStore) AddOpenQuestion(question string) {
	question = trimSpace(question)
	if question == "" {
		return
	}
	for _, q := range s.state.OpenQuestions {
		if q == question {
			return
		}
	}
	s.state.OpenQuestions = append(s.state.OpenQuestions, question)
	s.touch()
}

// Finish 收敛到终态。
func (s *AgentStateStore) Finish(status AgentRunStatus, reason AgentStopReason) {
	s.state.Status = status
	if reason != "" {
		s.state.StopReason = reason
	}
	s.touch()
}

func cloneState(src *AgentState) *AgentState {
	out := *src
	out.Plan = append([]AgentPlanStep{}, src.Plan...)
	out.CompletedSteps = append([]AgentPlanStep{}, src.CompletedSteps...)
	out.PendingSteps = append([]AgentPlanStep{}, src.PendingSteps...)
	out.LoadedSkills = append([]string{}, src.LoadedSkills...)
	out.Observations = append([]AgentObservation{}, src.Observations...)
	out.Evidence = append([]AgentEvidence{}, src.Evidence...)
	out.OpenQuestions = append([]string{}, src.OpenQuestions...)
	out.Assumptions = append([]string{}, src.Assumptions...)
	return &out
}
